src/utils/messages.js

```diff
@@ -1,20 +1,28 @@
-const generateMessage = (username, text) => {
-    return {
-        username,
-        text,
-        createdAt: new Date().getTime(),
-    }
-}
+const User = require('../entities/user')
 
-const generateLocationMessage = (username, url) => {
-    return {
-        username,
-        url,
-        createdAt: new Date().getTime(),
-    }
-}
+const admin = new User({ username: 'admin' })
+
+const makeMessageGenerator =
+    (author, process = (p) => p, payloadKey = 'text') =>
+    (payload) => ({
```

Comment:
Do I understand correctly that this means that the `makeMessageGenerator` function returns a function?
Could you please walk me through this way of writing functions?

I can't wrap my head around the `payload` parameter, I mean where does it come from if you've already passed stuff into the `makeMessageGenerator` ?
Why does this 'nameless' function has round brackets after the 'arrow' symbol?
Does this approach of writing functions have a name? I mean to read about it.

Reply:
```

Reply:
```
Do I understand correctly that this means that the makeMessageGenerator function returns a function?
```
Yes.

```
Could you please walk me through this way of writing functions?
```
https://medium.com/@iampika/part-6-javascript-functions-returning-functions-from-functions-429a3d9a55d1


```
I can't wrap my head around the payload parameter, I mean where does it come from if you've already passed stuff into the makeMessageGenerator ?
```
`makeMessageGenerator` returns a funtion which takes `payload` as an argument
```javasript
const createAdminMessage = makeMessageGenerator({username: 'Admin'}, (p) => p, 'text')
const message = createAdminMessage('Hi there!') // => payload === 'Hi there!' => true 
```


```
Why does this 'nameless' function has round brackets after the 'arrow' symbol?
```
Because JS interpreter does not understand is it a function body or an object in curly braces 



```
Or the bit where you add a property to the object named as the 'payloadKey' and you make it equal to the payload that you put through the passed process. I mean I think I mostly see what goes where, I just don't quite get 'why do it like this'?
``` 
This is the way to define dynamic attributes. I need this because different messages have different structure in the project (text and location) 

```
Does this approach of writing functions have a name? I mean to read about it.

Reply:
I'm scared to say it but I think I understand the approach. (fingers crossed)
However, can we get back to curly brackets for a sec?
```
Because JS interpreter does not understand is it a function body or an object in curly braces
```
Do you mean that the round brackets around the curly ones are "just for the interpretor"?
And in this case it doesn't really mean that we are returning another function? :/
It's just that in the suggested articles they didn't use brackets for the "final return".
`const sum = x => y => z => x + y + z`

Reply:
You are right